modules/plugin/mongodb/src/main/java/org/geotools/data/mongodb/AbstractFilterToMongo.java

```diff
@@ -520,27 +543,88 @@ public Object visit(Disjoint filter, Object extraData) {
         throw new UnsupportedOperationException();
     }
 
+    /**
+     * DWITHIN filter is delegated to Mongo's $near operator for point geometry, otherwise it is
+     * delegated to $geoIntersects. To properly compute the buffer the azimuthal equidistant
+     * projection is being used.
+     *
+     * @param filter
+     * @param extraData
+     * @return
+     */
     @Override
     public Object visit(DWithin filter, Object extraData) {
         BasicDBObject output = asDBObject(extraData);
 
         Object e1 = filter.getExpression1().accept(this, Geometry.class);
 
         Geometry geometry = filter.getExpression2().evaluate(null, Geometry.class);
-        double distanceInMeters = DistanceBufferUtil.getDistanceInMeters(filter);
 
-        DBObject geometryDBObject = geometryBuilder.toObject(geometry);
-        DBObject dbo =
-                BasicDBObjectBuilder.start()
-                        .push("$near")
-                        .add("$geometry", geometryDBObject)
-                        .add("$maxDistance", distanceInMeters)
-                        .get();
+        CoordinateReferenceSystem coordinateReferenceSystem =
+                featureType.getCoordinateReferenceSystem();
+        double distanceInMeters = DistanceBufferUtil.getDistanceInMeters(filter);
+        DBObject dbo;
+        // If point use $near
+        if (geometry instanceof Point) {
+            DBObject geometryDBObject = geometryBuilder.toObject(geometry);
+            dbo =
+                    BasicDBObjectBuilder.start()
+                            .push("$near")
+                            .add("$geometry", geometryDBObject)
+                            .add("$maxDistance", distanceInMeters)
+                            .get();
+            // Other type of geometry, use $geoIntersects
+        } else {
+            MathTransform transform = null;
+            try {
+                CoordinateReferenceSystem azimuthalEquidistantCRS =
+                        computeAzimutalEquidistantCRS(geometry, coordinateReferenceSystem);
+                // Transform geometry to CRS that supports meters
+                transform =
+                        CRS.findMathTransform(coordinateReferenceSystem, azimuthalEquidistantCRS);
+                geometry = JTS.transform(geometry, transform);
+
+                // Apply buffer in meters to geometry
+                Geometry bufferedGeometry = geometry.buffer(distanceInMeters, DEFAULT_SEGMENTS);
+
+                // Transform back to original with buffer applied
+                transform =
+                        CRS.findMathTransform(azimuthalEquidistantCRS, coordinateReferenceSystem);
+                geometry = JTS.transform(bufferedGeometry, transform);
+            } catch (FactoryException | TransformException e) {
+                throw new RuntimeException(e);
```

Comment:
Provide some context, e.g.:

```suggestion
                throw new RuntimeException("Failed to compute polygon within distance from reference geometry", e);
```

Reply:
Added message